Feature: share articles

mnassabain opened the pull request:
## 🚀 Features

This update contains the following features:

- Possibility to share an article on social media (Facebook, Twitter) or via mail
- Possibility to share an article with a fellow nextcloud user
- Allow users to view articles that are shared with them
- Multi-language support for the share functionality
- Tests validating the share functionality
- UI/UX friendly design

## 🖱️ UI/UX implementation

- Each article has now a share button next to the stare and read buttons

![image](https://user-images.githubusercontent.com/34754819/108606105-70e27c80-73b8-11eb-9f23-95f488729061.png)

- Clicking on the button opens a dropdown menu, allowing users to either share the article on social media, or search for a user and share with them. Clicking on a user shares the article with them.

![image](https://user-images.githubusercontent.com/34754819/108606110-7b9d1180-73b8-11eb-8c91-1eeda7882a17.png)

- Each user can view articles that were shared with them by clicking on the "Shared" menu. Each available article can be starred, marked as read/unread, and even shared with someone else

![image](https://user-images.githubusercontent.com/34754819/108606114-835cb600-73b8-11eb-9a32-027f958b331b.png)

## 💻 Code implementation

Sharing an article is implemented by duplicating it. Here's a detailed analysis of our decision.

### ❌ Idea 1: Sharing by reference

- Firstly, we have considered to share by reference, adding a table with three fields : `item_id`, `shared_by`, and `shared_with`.
- This idea avoids data redundancy
- However, other functionalities, like starring, or marking an item as read, wouldn't work. `starred` and `unread` are columns of the `news_items` model, making it a one-on-one relation. This means that a shared article also shares its' `starred` and `unread` state between users, which is something we don't want.
- We could implement the `starred` and `unread` fields into the share table, however, this would create a lot of code redundancy, since we would have to rewrite the same code, only for another object.

### ✅ Idea 2: Sharing by copying

- This idea seems to work better.
- Each item could be starred or read independently, since each user has his own copy.
- This would come at a cost of some data duplication, however, it would save us from code redundancy.
- Here's a step-by-step process of sharing an item:
    1. User "A" decides to share an item with user "B".
    2. We check if the share relation exists - if the article is already shared, we exit the function.
    3. If not, the function continues.
    4. We duplicate the item.
    5. We initialize the values of the duplicated item:
        - starred = false
        - unread = true
        - shared_by = "A"
        - shared_with = "B"
        - feed_id - same as the original article - a shared item is considered to be still "owned" by "A", and is in his feed. This allows us to track the origin of the shared article.
    6. The new item is inserted into the DB.
- We needed to modify some existing code in order for a user to access items shared with them. A couple of functions in the `ItemMapper` have been modified, mostly with the following update:

```php
// when fetching articles for a user
'AND ('.
      '(`feeds`.`user_id` = ? AND `items`.`shared_by` LIKE \'\')'.
      ' XOR `items`.`shared_with` = ?' .
 ') '
// either the article is in the users feed and not shared
// (remember, a shared item is still in sharers feed, however it's not really his)
// or it's not in his feed and is shared with him
```

### 🧪 Tests

All of our functions are tested to work properly, and existing tests were modified to take into account the share data.

## 👥 Team

We're a team of six people working on this project for our university. Here are the members:

- Marco NASSABAIN
- Nicolas WENDLING
- Jimmy HUYNH
- Aurélien DAVID
- Hamza ELHADDAD
- Ilyes CHERGUI-MALIH

We are from the University of Strasbourg. Special thanks for M. Julien GOSSA for giving us this subject and helping us getting started.

## 🗓️ Future updates

We are currently starting to work on the addition of three new functionalities for the next part of our project :

- Adding an excerpt of the article to the social media post
- Automatic hashtag finding for the social media post
- Allowing admins to share feeds with their group ([discussion](https://github.com/nextcloud/news/discussions/550))

## 🙏 Final words

We have been working on this project for a while, and we have thoroughly studied the existing code before contributing. We have tried our best to write good and understandable code. We have ran manual and automatic tests which the functionality passes.

However, it is possible that we have a couple of bugs lying around. We're open to critisicm, and we'd like to hear what you think about our work.

I hope you like our contribution, as much as we liked contributing! Thank you.

SMillerDev commented:
There is 207 files changed and there are code conflicts. Please split this review into at the very least API and frontend changes and make sure it's actually mergable.

Grotax commented:
First good step would be to remove all the language file changes, we don't need those and it makes the overview difficult. But I appreciate the effort.
https://github.com/nextcloud/news/blob/master/CONTRIBUTING.md#translation

mnassabain commented:
Thank you for the feedback.
I have started by removing all the translation files.

We have just realised that we are actually way behind your recent changes, so we will take some time to adapt our code to make it match your updates.

SMillerDev commented:
Like mentioned before:

> Please split this review into at the very least API and frontend changes [...]

In addition to that we'll need all commits to have a sign off to comply with the DCO requirement for the nextcloud organisation.

mnassabain commented:
Yup, still working on it.
It's taking a bit more time than expected because rebasing/cherry-picking requires us to re-resolve old merge conflicts, and we're trying to configure git to remember the resolves and replay them.

Should be done soon, I'll leave a comment when it's ready for review :)

mnassabain commented:
Ok, I think we're ready for review!
This pull request now contains only back-end changes (tests included, please don't ask me to split those too haha it's hard work)
Pull request #1217 contains changes to the front-end.

We've tried to write clear code, and comment parts that were a bit complicated. Also I've put some details about our implementation in the pr description. I'll be on watching for your questions or comments!

Thanks

SMillerDev commented on appinfo/routes.php:
```diff
@@ -50,6 +50,7 @@
 ['name' => 'item#read', 'url' => '/items/{itemId}/read', 'verb' => 'POST'],
 ['name' => 'item#read_multiple', 'url' => '/items/read/multiple', 'verb' => 'POST'],
 ['name' => 'item#star', 'url' => '/items/{feedId}/{guidHash}/star', 'verb' => 'POST'],
+['name' => 'item#share', 'url' => '/items/{itemId}/share/{shareWithId}', 'verb' => 'POST'],
```
Share with is pretty ambiguous. I'd prefer `shareRecipientId`

SMillerDev commented on lib/Db/Item.php:
```diff
@@ -63,6 +63,11 @@ class Item extends Entity implements IAPI, \JsonSerializable
     protected $unread = false;
     /** @var bool */
     protected $starred = false;
+    /** @var string */
+    protected $sharedBy = '';
+    /** @var string */
+    protected $sharedWith = '';
```
This should be null if not set. Same issue as with the routes too, is it shared with Facebook or with a person? We don't know based on this name.

anoymouserver replied:
This would also improve the SQL queries. `shared_by = \'\'` could become `shared_by IS NULL` which is much cleaner.

SMillerDev commented on lib/Db/ItemMapperV2.php:
```diff
@@ -56,9 +56,10 @@ public function findAllFromUser(string $userId, array $params = []): array
                 ->from($this->tableName, 'items')
                 ->innerJoin('items', FeedMapperV2::TABLE_NAME, 'feeds', 'items.feed_id = feeds.id')
-                ->where('feeds.user_id = :user_id')
+                ->where('(feeds.user_id = :user_id AND items.shared_by = \'\') OR items.shared_with = :shared_with')
```
I don't think shared articles should be in your normal list since we can't explain how they got there.

anoymouserver replied:
I agree with @SMillerDev on that one.

Apart from that, the query could probably be simplified
```sql
feeds.user_id = :user_id OR items.shared_with = :user_id
```
The `:user_id` parameter ca be reused as it contains the same value anyways .. and from what I've seen so far `shared_by` is always empty if the feed's user matches.

mnassabain replied:
We've thought about this and for us it makes sense to include shared articles in this function as well as in findAllItems, since they should be shown to the user.

Also, these functions are used in other functions which need access to shared items. For example, if we star an item that is shared with us, we want it to appear in the "Starred" menu. To retrieve the starred items we use _findAllItems()_, and the _starred()_ function used for the starred count uses _findAllFromUser()_

The simplification that you've suggested wouldn't work in our implementation.
```sql
feeds.user_id = :user_id OR items.shared_with = :user_id
```
Here's why: if user A shares an item with user B, the item is still in user A's feed, with _shared_by_ = B.
If we want to find all items for user A, we need to verify that _shared_by_ is empty, otherwise it's an article that's shared with someone. That article should instead be shown to user B.

SMillerDev commented on lib/Service/ItemServiceV2.php:
```diff
@@ -391,4 +429,62 @@ public function findAllWithFilters(
+        }
+
+        // duplicate the item
+        $sharedItem = Item::fromImport($item->jsonSerialize());
```
Why not `clone $item`

anoymouserver commented:
"**Sharing by copying**" could maybe also be done by creating a dummy feed which isn't updated (like the [import does](https://github.com/nextcloud/news/blob/7e3ddf4af023b32269c94328eecec1103f1c4f9c/lib/Service/ImportService.php#L78) for unknown items), which gets all the shared items assigned. This would prevent vanishing of the item if the original feed is deleted, would integrate seamlessly into the structure and would only require one additional feed to track who shared the item (`shared_by`). The drawbacks are that the item would not be updated by the feed fetcher (but that also the case for the current implementation) and every user who received a shared item would then have an additional feed somewhere.

(@SMillerDev correct me if I haven't considered something important)

SMillerDev commented:
No, that's a good idea. I also like this approach because it gives the receiving party some control over the sharing (they could remove the feed and everything in it).

It would also simplify ui and implementation.

mnassabain commented on lib/Service/ItemServiceV2.php:
```diff
@@ -391,4 +429,62 @@ public function findAllWithFilters(
+        $sharedItem->setSearchIndex($item->getSearchIndex());
+
+        // set share data
+        $sharedItem->setSharedBy($userId);
+        $sharedItem->setSharedWith($shareWithId);
```
In most cases yes, but not always.
Say that user A shares an article with user B. In this case it would work because the feed belongs to A.

However, if user B decides to share that same article with user C, there would be a problem. The owner of the article is still user A, so C would think that A shared it with them.

feed owner = article owner
shared_by = sender
shared_with = recipient

anoymouserver replied:
If allowing to re-share a shared article the long-term consideration gets even more important.

Following scenario: User1 shares with User2, who shares with User3 after 14 days ...
- U1 now deletes the feed after 14.5 days which removes all of the shared items but U3 only had like 12 hours to read it and knew that U2 shared it.
- U3 could now share with U1 since they didn't knew that it started from them (causing the item to be duplicated for U1)

mnassabain commented on lib/Service/ItemServiceV2.php:
```diff
@@ -391,4 +429,62 @@ public function findAllWithFilters(
+        // copy and initialize fields
+        $sharedItem->setUnread(true);
+        $sharedItem->setStarred(false);
+        $sharedItem->setFeedId($item->getFeedId());
```
Yes, we did think about that, but we didn't really see that as a problem.
The way we've visualised it is that, unlike sharing feeds or folders, which would be constant, sharing an article is more of an instantanious action. The information is transmitted, and once the article is read, it's marked as such and hidden (if not starred)
Same thing here, we don't think of the share to be something long term, so we don't think it's that big of a problem

anoymouserver replied:
But isn't starring the article something long-term? At least I considered it this way.

mnassabain commented:
I agree that that is a good idea, we can implement it that way.
We've already thought this option, but we didn't really like the idea of creating dummy feeds.
It does seem to be easier to implement, and it would even simplify some front-end code.

If that's the way you prefer it done, we could look into it?

anoymouserver commented:
Well, I don't want to force anything on you, but I see some side effects with your approach.

1. the sharing user decides that the subscribed feed is no longer wanted, which would also delete every shared item, regardless of whether they are unread, starred or new.
2. the purging, which deleted old read items to save space, skips the top n items per feed (since these are the new ones) and removes the other ones (keeping at least `articles_per_update` items to avoid refetching).
   If a user now shares 50 items (exaggerated to illustrate) it would shift the most recent fetched items by 50, which would delete some items which should be kept for now.
3. if the feed fetcher finds updated but already existing items, it updates them in the database. This does not work correctly if the item exists multiple times per feed.

If you have a better suggestion on how to get around these issues, go for it. This was the simplest approach I could think of with my knowledge of the codebase.

mnassabain commented:
Yes I see.. it's true that those problems exist, and while we can fix them by changing more code, I think that your solution is more elegant. We'd remove a lot of our logic from already existing functions, and the dummy feed would already be handled by existing code.
It's a cleaner solution.

I'll look into it and play around with the code, see if any problems come up. I'll discuss it tomorrow with the team, and if everything is ok we'll push some code soon. Thanks for the help!

mnassabain commented:
I've modified the share item code to use dummy feeds.
When an item is shared, a feed belonging to the recipient is created, if it doesn't already exist. Shared items are then added to this feed, and _shared_by_ is set to contain the sender's id.
This update allowed us to remove a lot of share logic from the ItemMapper functions and to remove the _shared_with_ column.

SMillerDev commented on tests/Unit/Service/ShareServiceTest.php:
```diff
@@ -0,0 +1,216 @@
+            ->setSharedBy($this->uid);  // ... and contains the senders user ID
+
+        // Dummy feed 'Shared with me'
+        $feed = new Feed();
```
This should be a mock to make sure it's setting the correct values.

mnassabain replied:
Cannot mock it because we use `new Feed()`, and I cannot test it by testing the argument of the `feedService->insert` function - because the time field will be variable. I don't see a solution to test this, that's why the first time I left it like this

anoymouserver commented on lib/Db/Item.php:
```diff
@@ -482,6 +516,16 @@ public function setTitle(string $title = null): self
         return $this;
     }
 
+    public function setSharedBy(string $sharedBy = null): self
```
Sure about that? It's exactly as every other `string` setter in here. 🤔

SMillerDev replied:
That's worse. I'll check.

anoymouserver commented on lib/Service/ShareService.php:
```diff
@@ -0,0 +1,112 @@
+                 ->setUrlHash(md5($feedUrl))
+                 ->setLink($feedUrl)
+                 ->setUrl($feedUrl)
+                 ->setTitle('Shared with me')
```
It would also vote for [translation support](https://docs.nextcloud.com/server/latest/developer_manual/basics/front-end/l10n.html#php) here (I know the import currently doesn't).

mnassabain replied:
Just to be sure, localized as in writing a function in the feedService that fetches/creates a dummy url?
I've thought about doing this, but since I was basing my code on the ImportService, I thought it would be better to follow in your footsteps.
Good idea for the translation support, I will add that in.

anoymouserver replied:
As in the following, plus the `IL10N` in `__construct`.
```suggestion
                 ->setTitle($this->l->t('Shared with me'))
```

SMillerDev replied:
Localization is what you call translation support. Since you're not actually changing anything for the locale I guess both are valid: https://lokalise.com/blog/translation-and-localization-difference/

mnassabain replied:
Oh yea 🤦‍♂️😅 my bad

anoymouserver commented on lib/Service/ShareService.php:
```diff
@@ -0,0 +1,112 @@
+        $sharedItem->setSharedBy($userId);
+
+        // Get 'shared with me' dummy feed
+        $feedUrl = 'http://nextcloud/sharedwithme';
```
I'm not sure how well a complete dummy url works, but the import uses the same pattern (`http://nextcloud/nofeed`).
Maybe something standardized like `urn:nextcloud:sharedwithme` would work too but I'm not sure whether clients expect an actual http(s) url.

SMillerDev replied:
On that case I'd go for `https://serverurl/news/shared` or something. Arguably we can get away with a lot more magic on imports because the user initiates it.

anoymouserver commented on lib/Service/ShareService.php:
```diff
@@ -0,0 +1,112 @@
+
+        // Duplicate item & initialize fields
+        $sharedItem = clone $item;
+        $sharedItem->setId(null);
```
MySQL/MariaDB and SQLite would use the auto-increment value if `NULL` is provided, but I don't know the doctrine handling of it.
And the ID of the cloned item has to be reset for sure.

mnassabain replied:
It wasn't valid at first, I had to implement a __clone() method in Item that, in addition to cloning the element, marks all the feeds as updated. This was necessary because the insert() function only inserts columns that are marked as updated.
I didn't know whether I should put the setId(null) in that function or not, but we think it's better outside since a clone should arguably have the same id as the cloned item.

Anyways, I've tested the api call and it works, the item is inserted with a new id

anoymouserver commented:
Looks quite good so far, I would test it on my end as soon as the frontend is ready.

SMillerDev commented on lib/Service/ShareService.php:
```diff
@@ -44,20 +46,36 @@ class ShareService
     public function __construct(
         FeedServiceV2 $feedService,
         ItemServiceV2 $itemService,
+        IURLGenerator $url,
+        IL10N $l,
```
please give these more descriptive names. It's not a URL. It's a URL generator.

mnassabain replied:
My bad, I was basing my implementation on existing code (in Settings/AdminSection, `IURLGenerator $url`)
I've corrected the ShareService code.

By the way, could we get an update on the unresolved reviews? We've resolved some of them, but there are a couple that are still open:
https://github.com/nextcloud/news/pull/1191#discussion_r588856418
https://github.com/nextcloud/news/pull/1191#discussion_r588855739
https://github.com/nextcloud/news/pull/1191#discussion_r588859055

Thanks!

SMillerDev commented:
Can you rebase this?

mnassabain commented:
I'm currently solving the merge commits. I don't understand what should I rebase it on? Can you clarify? Thanks!
Also, I have another quick question, should I keep the migration from this pr separate from the categories one? Or should I put both in one file? Since it's the same version 150302? Not sure what's the correct way and I don't want to break anything.

Thanks

SMillerDev commented:
I prefer a rebase flow where you run `git rebase [-i] origin/master` instead of merging master into your branch to get it up to date.

anoymouserver commented:
Rebasing means solving the merge conflicts by using rebase and then force-pushing instead of a merge commit.

Regarding the migration php I also thought about combining them, but the 150302 should probably be a 150400 since the 15.3.2 ist already released.

mnassabain commented:
Oh all right, thanks, I'll get onto that asap.

Meanwhile, we've gotten a better idea for the sharedBy field. Currently it's a string containing the user id of the sharer. It would be better if it was a **foreign key** referencing the sharer. In my opinion t's a far better solution, I don't know why we didn't think of it earlier..

SMillerDev commented:
Since we don't control the user table I'm not sure that's such a great idea.

mnassabain commented:
Right, I didn't take that into consideration.. I guess there are some faults with that approach then.. In that case we'll stick with what we've got

mnassabain commented:
Added something new before rebasing.. since it's useful to have the sharers display name in the front-end, I've updated the model and mapper to fetch them.

My method is the following:
1. I've added a field into the Item model `sharedByDisplayName` (maybe not the best name) with a setter and a getter - this field is not in the database, it's only used by the mapper
2. I've added a findEntities function into the Item mapper, that firstly, calls the corresponding parent function, fetching the entities, and then loops over them, fetches the sharer display names, and adds them into the `sharedByDisplayName` field.
3. I've also updated the tests to take these changes into account.

This is the simplest solution I could find, let me know what you think! Thanks!

mnassabain commented:
I've rebased this pr onto master like asked.
I've also implemented a way to fetch sharer display names, I've described it in [this comment](https://github.com/nextcloud/news/pull/1191#issuecomment-802322362), I'm mentioning it here just in case the comment gets lost.

Looking forward to hearing your thoughts! Thanks

SMillerDev commented on lib/Controller/ItemController.php:
```diff
@@ -318,4 +325,26 @@ public function readMultiple(array $itemIds): void
+     * @param int $itemId              Item to share
+     * @param string $shareRecipientId User to share the item with
+     */
+    public function share($itemId, $shareRecipientId)
```
```suggestion
    public function share(int $itemId, string $shareRecipientId)
```

mnassabain commented on lib/Db/ItemMapperV2.php:
```diff
@@ -31,15 +32,50 @@ class ItemMapperV2 extends NewsMapperV2
+     * @var IUserManager
+     */
+    private $userManager;
+
```
That makes sense, I agree.. I've picked this method because it allowed me to modify the least amount of code.

There's another option that I've thought of, and I've now committed it. I moved the display name mapping into a function in the ShareService. I call this function from the ItemController when returning fetched items.

Is that better?

SMillerDev replied:
Sounds good

mnassabain replied:
Great, everything is pushed! Resolving this comment

anoymouserver reviewed (approved):
Looks good to me.

mnassabain commented:
Hello, just wanted to check if there are any updates regarding the review of this feature? We're available for any questions or code changes.

Thanks!

SMillerDev commented:
Just want to get 15.4 released before I merge this. But maybe we can start a stable-15 branch for that.

mnassabain commented:
Perfect, thanks! Let me know if there's anything you need me to do

SMillerDev commented:
@mnassabain can you rebase this?

mnassabain commented:
Done!

SMillerDev commented:
https://github.com/nextcloud/news/issues/1375
